### PR TITLE
CUBLAS: wrap rot! to implement rotate! and reflect!

### PR DESCRIPTION
**Pull request by maleadt**
https://github.com/JuliaGPU/CUDA.jl/pull/419

### CODE REVIEW

**Comment by maleadt** [comment]
@amontoison it fails with one combination of complex inputs though, see the `@test_broken`. Could you have a look?

**Comment by amontoison** [comment]
Cosines `c` can't be complex numbers : https://docs.nvidia.com/cuda/cublas/index.html#cublas-lt-t-gt-rot
`rotate!` updates in-place `x` and `y` such that
```julia
[x] = [   c     s] [x]
[y]   [-conj(s) c] [y]
```
whereas `reflect!` does the update
```julia
[x] = [   c     s] [x]
[y]   [conj(s) -c] [y] 
```
If we want to make them working with complex sines `s`, we need to update the `reflect!` function a little bit.

File: lib/cublas/linalg.jl [hunk 1]
```diff
@@ -47,10 +48,26 @@ function LinearAlgebra.axpy!(alpha::Number, x::CuArray{T}, y::CuArray{T}) where
...
+    nx==ny || throw(DimensionMismatch("rotate arguments have lengths $nx and $ny"))
+    rot!(nx, x, 1, y, 1, c, s)
+end
+
+function LinearAlgebra.reflect!(x::CuArray{T}, y::CuArray{T}, c, s) where T<:CublasFloat
+    nx = length(x)
+    ny = length(y)
+    nx==ny || throw(DimensionMismatch("reflect arguments have lengths $nx and $ny"))
+    rot!(nx, y, 1, x, 1, s, c)
+    swap!(nx, x, 1, y, 1)
```
Comment (amontoison) [inline lib/cublas/linalg.jl]:
```julia
rot!(nx, x, 1, y, 1, c, s)
scal!(ny, -real(one(T)), y, 1)
x, y
```

Comment (amontoison) [reply lib/cublas/linalg.jl]: It should work with complex sines `s`. `scal!` is also less expensive than a `swap!`.

Comment (maleadt) [reply lib/cublas/linalg.jl]: Thanks!

**Pull request merged**